### ISSUE TITLE
DIB renderer is really slow when back propagating

### ISSUE THREAD

**Issue by MSKTHU** [NVIDIAGameWorks/kaolin#237]
I am doing 3D reconstruction learning and found most of learning time costs on the DIB vertexColor renderer backpropagation.
Here is my test code, where shape, color and tri are from a 3D face morphable model. By the way, I change the code of DIB vertexColor renderer so that delta is a controllable parameter from outside.
```
import torch
from kaolin.graphics import DIBRenderer as Renderer
import os
import imageio
import numpy as np

os.environ["CUDA_VISIBLE_DEVICES"] = "1"

if __name__ == "__main__":

    shape = torch.from_numpy(np.load("./dump/shape.npy")).float().requires_grad_().cuda()
    color = torch.from_numpy(np.load("./dump/color.npy")).float().requires_grad_().cuda()
    tri = torch.from_numpy(np.load("./dump/tri.npy")).float().requires_grad_().cuda()

    renderer = Renderer(256, 256, mode="VertexColor")

    eyeMtx = torch.eye(3, 3).repeat(color.size()[0],1,1).cuda()
    zeroVec = torch.zeros(3).repeat(color.size()[0],1).cuda()
    matrixP = torch.Tensor([2.0, 2.0, -1.0]).repeat(color.size()[0],1).cuda()
    renderer.set_camera_parameters([eyeMtx, zeroVec, matrixP])
    
    import time
    for i in range(20000):
        start = time.time()
        renderedAttr, renderedProb, _ = renderer(points = [shape, tri.long()], colors_bxpx3 = color, delta = 1000)
        loss = torch.sum(renderedAttr)
        print("time 1:",time.time() - start)
        start = time.time()
        loss.backward()
        print("time 4:",time.time() - start)
```
The results are listed as follow:
```
time 1: 0.0012078285217285156
inside rasterizer bp time 2: 3.814697265625e-05
inside rasterizer bp time 3: 2.9087066650390625e-05
time 4: 0.7192938327789307

time 1: 0.0011813640594482422
inside rasterizer bp time 2: 0.0002818107604980469
inside rasterizer bp time 3: 2.4080276489257812e-05
time 4: 0.7189295291900635
```
The inside raterizer bp time is test inside the LinearRasterizer.backward function of the file /kaolin-master/kaolin/graphics/dib_renderer/rasterizer/rasterizer.py: line 187:
```
        start = time.time()
        dldp2 = torch.zeros_like(tfpoints2dmul_bxfx6)
        dldp2_prob = torch.zeros_like(tfpoints2dmul_bxfx6)
        dldc = torch.zeros_like(tfcolors_bxfx3d)
        print("inside rasterizer bp time 2:", time.time() - start)
        start = time.time()
        cuda_rasterizer.backward(dldI_bxhxwxd.contiguous(),
                                 dldp_bxhxwx1.contiguous(),
                                 tfims_bxhxwxd, tfimprob_bxhxwx1,
                                 tfimidxs_bxhxwx1, tfimweis_bxhxwx3,
                                 tfprobface, tfprobcase, tfprobdis, tfprobdep, tfprobacc,
                                 tfpoints2dmul_bxfx6, tfcolors_bxfx3d,
                                 dldp2, dldc, dldp2_prob,
                                 debug_im, multiplier, delta)
        print("inside rasterizer bp time 3:", time.time() - start)
```
As you can see, the computing time inside the backward method is really fast, which is 1e-5 second. However, the total backpropagating time is 0.7 second. 
Could you help me with this issue? Thanks

I am using Pytorch 1.4.0 + CUDA 10.1 + Cudnn 7.6.5 + python 3.7 + Ubuntu16.04 + Titan XP
The batch size is 3
The number of vertices is 37049 and number of triangle faces is around 70700

**Comment by ebartrum** [comment]
Would there be any way for me to speed this up since in my use case I am only optimising the texture and not the shape? I.e vertices are fixed. I'm using VCRenderer.

**Comment by MSKTHU** [comment]
> Would there be any way for me to speed this up since in my use case I am only optimising the texture and not the shape? I.e vertices are fixed. I'm using VCRenderer.

I think the renderer's back propagation will not speed up since the gradients of shape will be computed no matter if you will use the gradients or not. You can try.

**Comment by Caenorst** [comment]
Closing for inactivity, DIB-R have changed a lot in the last year.
